Flatiron - Tuples chapter.md

```diff
@@ -0,0 +1,115 @@
+let simpleDictionary = [ "Darth":"big baddie", "Ben":"mentor", "Luke":"protagonist" ]
+let richerDictionary = [
+	"Darth":[ "big baddie", "father of Luke" ],
+	"Ben":[ "mentor", "trainer of Darth" ],
+	"Luke":[ "protagonist", "son of Darth" ]
+	]
+print( richerDictionary[ "Darth" ] )
+[ "big baddie", "father of Luke" ]
+print( richerDictionary[ "Darth" ]![1] )
+"father of Luke"
+```
+Dicitionaries are very useful when you want a natural way of accessing contents and, crucially, you have a way of uniquely identifying each entry.
+
+**Classes and Structs** You may have already heard about these but they haven't yet been covered in class.  In brief, classes and structs are a powerful way to represent data because they guarantee the presence of specific properties (kind of like a dictionary which guarantees that there are entries for specific keys and, furthermore, can't have any more or fewer entries).  Classes and Structs are great to model specific types of data where you know exactly what should be present, such as a "Student" having a first name, last name, and array of test grades.
+
+**Tuples** are a special kind of tool that combine properties of arrays, dictionaries, and classes but are not interchangeable with any of them.  Yup, you read that right, they're kind of like a Franken-data structure.  
+Like an array tuples allow you to access their contents by index.  
+Like a dictionary tuples allow you to access their contents by name (if you specify names). 
+Like a class or struct tuples have *exactly* the contents described in their declaration and of the type indicated in their declaration.  (This is checked for you by the compiler.  Remember, one of the key aspects of Swift is type-safety).
+
+Arrays and Dictionaries are open-ended: they can contain some elements, thousands of elements or no elements, and the contents can grow or shrink.  They're the right solution when you don't know exactly what you have and need to iterate over the contents to search for an item or do something to each element.
+
+Classes, Structs and Tuples are closed: once they've been declared the compiler will guarantee you exactly what they.  They're the right solution when you want to represent structured data and get help from the compiler to make sure its correct.
+
+##What's a tuple good for?
+Now you're probably wondering why on earth you would want to use a tuple and, if that's where your mind has gone, I can't blame you.  Rest assured that there *are* some circumstances where a tuple is extremely useful, kind of like that weird tool you walk by in the hardware store that twists your mind but, in the right hands and in the right situation, can get the job done like nothing else.
+
+Tuples are the tool of choice when you want to pass highly structured data *to* a function or return highly structured data *from* a function but don't need the additional capabilities of a full-blown class or struct and don't want to go through the trouble of creating one.
+
```

Comment:
This is a great intro. I think right after you present something like this, you can jump right into showing the student what a tuple looks like. Easing them into seeing the most basic form of a tuple, its components and how it represents what it is you stated here.

```swift
let numberInfo = (min: 1, max: 105)
```

_Here we can break down what it is this thing is_

Well, what does this all mean? Lets give you an example. If we have an array of `Int`'s, and we want to create a function that iterates over the supplied array and returns back to us the minimum value, can you do that? Sure.

```swift
func minNumber(numbers: [Int]) -> Int {
    
    // do work to figure out the min
    // assuming 1 was found to be the min
    
    return 1
}
```

But, what if wanted to see the max as well?